gcc/config/riscv/riscv.c

```diff
@@ -4200,26 +4203,15 @@ riscv_trampoline_init (rtx m_tramp, tree fndecl, rtx chain_value)
   emit_insn (gen_clear_cache (addr, end_addr));
 }
 
-/* Return leaf_function_p () and memoize the result.  */
-
-static bool
-riscv_leaf_function_p (void)
-{
-  if (cfun->machine->is_leaf == 0)
-    cfun->machine->is_leaf = leaf_function_p () ? 1 : -1;
-
-  return cfun->machine->is_leaf > 0;
-}
-
 /* Implement TARGET_FUNCTION_OK_FOR_SIBCALL.  */
 
 static bool
 riscv_function_ok_for_sibcall (tree decl ATTRIBUTE_UNUSED,
 			       tree exp ATTRIBUTE_UNUSED)
 {
-  /* When optimzing for size, don't use sibcalls in non-leaf routines */
+  /* When save-restore routine, don't use sibcalls in non-leaf routines */
   if (TARGET_SAVE_RESTORE)
-    return riscv_leaf_function_p ();
+    return cfun->machine->use_save_libcall_p;
```

Comment:
The logic here is backwards.  It should be !cfun->machine->use_save_libcall_p, since we want to disable sibcalls if a save routine was used in the prologue.

The reason your patch works is because the ok_for_sibcall hook is called when we lower IL from gimple to RTL, and at this time use_save_libcall_p is always zero.  It doesn't get set until later when we expand prologues, which is one of the late RTL optimization passes.  Your patch is hence always returning false from ok_for_sibcall hook if TARGET_SAVE_RESTORE is true.  This does appear to be the correct simple solution, but it should be written directly that way.  So the use_save_libcall_p variable should be eliminated, and we should just return false here.

Reply:
Thanks for dig out that, I'll update patch tomorrow :)